File: ruby/01/06.rb
# -*- coding: utf-8 -*-

=begin
06. 集合
"paraparaparadise"と"paragraph"に含まれる文字bi-gramの集合を，それぞれ, XとYとして求め，XとYの和集合，積集合，差集合を求めよ．さらに，'se'というbi-gramがXおよびYに含まれるかどうかを調べよ．
=end

module Enumerable
  def ngram(n)
    each_cons(n).to_a
  end
end

words = %w[paraparaparadise paragraph]
x = words[0].gsub(' ', '').chars.ngram(2)
y = words[1].gsub(' ', '').chars.ngram(2)
  
puts "union: #{x | y}"
puts "intersection: #{x & y}"
puts "difference: #{x - y}"
puts "X includes 'se'?: #{x.include?(%w[s e])}"
puts "Y includes 'se'?: #{y.include?(%w[s e])}"

File: python/01/06.py
# -*- coding: utf-8 -*-
# 06. 集合
# "paraparaparadise"と"paragraph"に含まれる文字bi-gramの集合を，それぞれ, XとYとして求め，XとYの和集合，積集合，差集合を求めよ．さらに，'se'というbi-gramがXおよびYに含まれるかどうかを調べよ．
sentence1 = 'paraparaparadise'
sentence2 = 'paragraph'

def ngram(input, n):
    l = len(input) - n + 1
    list = []
    for i in range(0, l):
        list.append(input[i:i+n])
    return list

X = set(ngram(sentence1, 2))
Y = set(ngram(sentence2, 2))

print('union: {0}'.format(list(X.union(Y))))
print('intersection: {0}'.format(list(X.intersection(Y))))
print('difference: {0}'.format(list(X.difference(Y))))
print("X includes 'se'?: {0}".format('se' in X))
print("Y includes 'se'?: {0}".format('se' in Y))
